python/ray/tune/cloud.py

```diff
@@ -0,0 +1,316 @@
+import os
+import shutil
+import subprocess
+import tempfile
+from numbers import Number
+from typing import Optional, Dict
+
+from ray import logger
+from ray.tune.sync_client import (S3_PREFIX, GS_PREFIX, HDFS_PREFIX,
+                                  ALLOWED_REMOTE_PREFIXES)
+from ray.util import PublicAPI
+
+
+def is_cloud_target(target: str):
+    return any(target.startswith(prefix) for prefix in ALLOWED_REMOTE_PREFIXES)
+
+
+def _clear_bucket(bucket: str):
+    if not is_cloud_target(bucket):
+        raise ValueError(
+            f"Could not clear bucket contents: "
+            f"Bucket `{bucket}` is not a valid or supported cloud target.")
+
+    try:
+        if bucket.startswith(S3_PREFIX):
+            subprocess.check_call(
+                ["aws", "s3", "rm", "--recursive", "--quiet", bucket])
+        elif bucket.startswith(GS_PREFIX):
+            subprocess.check_call(["gsutil", "-m", "rm", "-f", "-r", bucket])
+        elif bucket.startswith(HDFS_PREFIX):
+            subprocess.check_call(["hdfs", "dfs", "-rm", "-r", bucket])
+
+    except Exception as e:
+        logger.warning(
+            f"Caught exception when clearing bucket `{bucket}`: {e}")
+
+
+def _download_from_bucket(bucket: str, local_path: str):
+    if not is_cloud_target(bucket):
+        raise ValueError(
+            f"Could not download from bucket: "
+            f"Bucket `{bucket}` is not a valid or supported cloud target.")
+
+    if bucket.startswith(S3_PREFIX):
+        subprocess.check_call(
+            ["aws", "s3", "cp", "--recursive", "--quiet", bucket, local_path])
+    elif bucket.startswith(GS_PREFIX):
+        subprocess.check_call(["gsutil", "-m", "cp", "-r", bucket, local_path])
+    elif bucket.startswith(HDFS_PREFIX):
+        subprocess.check_call(["hdfs", "dfs", "-get", bucket, local_path])
+
+
+def _upload_to_bucket(bucket: str, local_path: str):
+    if not is_cloud_target(bucket):
+        raise ValueError(
+            f"Could not download from bucket: "
+            f"Bucket `{bucket}` is not a valid or supported cloud target.")
+
+    if bucket.startswith(S3_PREFIX):
+        subprocess.check_call(
+            ["aws", "s3", "cp", "--recursive", "--quiet", local_path, bucket])
+    elif bucket.startswith(GS_PREFIX):
+        subprocess.check_call(["gsutil", "-m", "cp", "-r", local_path, bucket])
+    elif bucket.startswith(HDFS_PREFIX):
+        subprocess.check_call(["hdfs", "dfs", "-put", local_path, bucket])
+
+
+@PublicAPI(stability="beta")
+class TrialCheckpoint:
```

Comment:
I really like this idea of pairing local path and cloud path. Ideally I think things should be organized as:
on cloud: cloud_path_prefix  + common_upload_path
local: local_path_prefix + common_upload_path

We can use this to simplify the logic across the board.

One thing I am not so sure about is  why do we need to keep track of `self.metrics` here? This seems to be overly coupled with ExperiementAnalysis and will make this TrialCheckpoint class less generic.

Reply:
The initial idea for metric was to sometime extend this to a full `last_results` dict or so, so that `analysis.best_checkpoint.last_results` could be examined.

I'll get rid of that for now though, we can add this if there's demand for it.

Reply:
+1 to the prefix / common path abstraction. Let's go with the current replacement logic for now, but when refactoring the tech debt here we should do this!